File: backend/app/api/routes/coupons.py
"""Customer coupon API routes — check a code before checking out."""

from typing import Annotated

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.exceptions import AppError, raise_http_for_app_error
from app.core.rate_limit import rate_limit
from app.db.session import get_db
from app.dependencies.auth import CurrentUser
from app.dependencies.features import requires_feature_for_user
from app.schemas.coupon import CouponPreviewRequest, CouponPreviewResponse
from app.services.coupon_service import CouponService

router = APIRouter(prefix="/coupons")


def get_coupon_service(db: Annotated[Session, Depends(get_db)]) -> CouponService:
    return CouponService(db)


@router.post(
    "/preview",
    response_model=CouponPreviewResponse,
    dependencies=[Depends(requires_feature_for_user("coupons")), rate_limit("coupon_preview", 30, 300)],
)
def preview_coupon(
    data: CouponPreviewRequest,
    user: CurrentUser,
    service: Annotated[CouponService, Depends(get_coupon_service)],
) -> CouponPreviewResponse:
    """Advisory: shows what a code is worth against the cart the customer is looking at. The binding
    discount is worked out again from the server's own subtotal when the order is placed.

    Rate limited so the endpoint can't be used to guess codes at speed.
    """
    try:
        coupon, discount = service.preview(user.restaurant_id, data.code, data.subtotal, user)
    except AppError as exc:
        raise raise_http_for_app_error(exc) from exc
    return CouponPreviewResponse(code=coupon.code, description=coupon.description, discount=discount)

File: frontend/app/services/coupons.ts
import type { Coupon, CouponCreatePayload, CouponPreview } from '~/types/coupon'
import { apiFetch } from '~/services/http'

// ---- signed-in customer ---------------------------------------------------------------------

/** What a code is worth against the current cart. Advisory: the order endpoint decides for real. */
export function previewCoupon(code: string, subtotal: number) {
  return apiFetch<CouponPreview>('/api/v1/coupons/preview', {
    method: 'POST',
    body: JSON.stringify({ code, subtotal: subtotal.toFixed(2) }),
  })
}

// ---- admin ----------------------------------------------------------------------------------

export function fetchCoupons(restaurantId: number) {
  return apiFetch<Coupon[]>(`/api/v1/admin/coupons?restaurant_id=${restaurantId}`)
}

export function createCoupon(restaurantId: number, payload: CouponCreatePayload) {
  return apiFetch<Coupon>(`/api/v1/admin/coupons?restaurant_id=${restaurantId}`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function updateCoupon(restaurantId: number, couponId: number, payload: Partial<Coupon>) {
  return apiFetch<Coupon>(`/api/v1/admin/coupons/${couponId}?restaurant_id=${restaurantId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  })
}

export function deleteCoupon(restaurantId: number, couponId: number) {
  return apiFetch<void>(`/api/v1/admin/coupons/${couponId}?restaurant_id=${restaurantId}`, {
    method: 'DELETE',
  })
}
